Move release between folders / Remove from all folders

JOJ0 opened the pull request:
Fixes: #161 

- [x] New collection folder methods `move_release` and `uncategorize_release` using this API which description is very misleading but it is able to change the collection folder of a collection item: https://www.discogs.com/developers#page:user-collection,header:user-collection-change-rating-of-release
- [x] Added a test for it.
- [x] Updated documentation.

JOJ0 commented:
Yo @AnssiAhola, I'm desperately trying to mock to test a new feature and fail to mock this api: https://www.discogs.com/developers#page:user-collection,header:user-collection-change-rating-of-release

Help! I left some debug messages in the models code as well as in the test. If you happen to find the time to check this out, that would be awesome. Thanks!!!!

JOJ0 commented on discogs_client/models.py:
```diff
@@ -754,14 +757,38 @@ def releases(self):
+        self.client._post(
+            move_url,
+            {"folder_id": target_folder_id},
+        )
```
This code seems to work. I tried it out on the real API and I can successfully change folders of existing collection items.

JOJ0 commented on discogs_client/tests/test_models.py:
```diff
@@ -1,5 +1,5 @@
 import unittest
-from discogs_client.models import Artist, Release, ListItem, CollectionValue
+from discogs_client.models import Artist, Release, ListItem, CollectionValue, CollectionItemInstance
```
Since I need to pass a whole `CollectionItemInstance` to the new `move_release` method I import it here.

JOJ0 commented on discogs_client/tests/test_models.py:
```diff
@@ -314,6 +313,61 @@ def test_collection(self):
+        """Collection items can be moved to another folder"""
+        # Fetch the users collection folders from the filesystem
+        u = self.d.user("example")
+        self.assertEqual(u.collection_folders[2].id, 2)
```
This is actually redundant. It uses the filesystemfetcher. can be removed I think.

JOJ0 commented on discogs_client/tests/test_models.py:
```diff
@@ -314,6 +313,61 @@ def test_collection(self):
+            "/users/example/collection/folders/1/releases/123456": (b'{"instance_id": 123}', 201),
+
+            # Mocking the response of the POST to the instance resource URL
+            "/users/example/collection/folders/1/releases/123456/instances/123": (b"", 204),
```
This line should mock the response of when POSTing to the instance URL (see api docs link posted before). It returns a `204` and no data if I understand correctly.

JOJ0 commented on discogs_client/tests/test_models.py:
```diff
@@ -314,6 +313,61 @@ def test_collection(self):
+            #     },
+            #     "previous_request": None,
+            # },
+        )
```
Here i try to instantiate a minimalistic `CollectionItemInstance` so `move_release` can build the correct URL of it. I think id and instance_id should be enough for that purpose, right?

JOJ0 commented on discogs_client/tests/test_models.py:
```diff
@@ -314,6 +313,61 @@ def test_collection(self):
+        )
+
+        # Perform the move
+        u.collection_folders[1].move_release(instance, 2)
```
I get a 404 here:

```
self = <discogs_client.client.Client object at 0x103aacf80>, method = 'POST', url = ('/users/example/collection/folders/1/releases/123456/instances/123',), data = {'folder_id': 2}

    def _request(self, method, url, data=None):
        if self.verbose:
            print(' '.join((method, url)))

        self._check_user_agent()

        headers = {
            'Accept-Encoding': 'gzip',
            'User-Agent': self.user_agent,
        }

        if data:
            headers['Content-Type'] = 'application/json'

        content, status_code = self._fetcher.fetch(self, method, url, data=data, headers=headers)

        if status_code == 204:
            return None

        body = json.loads(content)

        if 200 <= status_code < 300:
            return body
        else:
>           raise HTTPError(body['message'], status_code)
E           discogs_client.exceptions.HTTPError: 404: Resource not found.
```

JOJ0 reviewed (commented):
Some inline comments so you understand what I'm thinking. Thanks!!!

AnssiAhola commented on discogs_client/models.py:
```diff
@@ -754,14 +757,38 @@ def releases(self):
+        if not isinstance(instance, CollectionItemInstance):
+            raise TypeError('instance must be of type CollectionItemInstance')
+        resource_url = self.fetch('resource_url')
+        move_url = f"{resource_url}/releases/{instance.id}/instances/{instance.instance_id}",
```
@JOJ0 I forgot to submit my review, so not sure if this comment was visible to you or not :sweat_smile: 
Actually responded 2 days ago, my bad.

JOJ0 replied:
Thanks so much @AnssiAhola. Fun or not so fun fact is: I recently did exactly the same on my day job: Copy/paste something, missing a comma, and searching for ages for a bug caused by a thing beinging a tuple instead of a string. Damnit! Thanks! 🤣

JOJ0 commented:
Ready for a  final review. Nitpicks on code, docs wording and example-code appreciated! Also I did some slight refactoring in CollectionFolders class for more modern Python and activated a Sphinx-MyST-plugin that allows creating "info boxes", "attention boxes" and  so on. You might like that @prcutler :-)

I rendered this branches docs changes for easier reviewing here: https://python3-discogs-client.readthedocs.io/en/foldermove/fetching_data.html#adding-a-release-to-a-collection-folder

Thanks in advance

JOJ0 commented on docs/source/conf.py:
```diff
@@ -80,7 +80,8 @@
 myst_enable_extensions = [
-    "substitution"
+    "substitution",
+    "colon_fence",
```
that is btw the extension used for info-boxes and so on @prcutler 🤩

JOJ0 commented on docs/source/fetching_data.md:
```diff
@@ -242,6 +242,10 @@ releases = folder.releases
+:::{caution}
+The {meth}`.remove_release` method deletes from the collection entirely. To remove an instance from a folder only, use the {meth}`.uncategorize_release` method.
+:::
+
```
and that's how u use colon_fence boxes

prcutler reviewed (approved):
Everything looks good to me, sorry for the delay.

JOJ0 commented:
@AnssiAhola I suppose no objections on your end since you had a look at the tests and code earlier already. I'll merge and release this now.

JOJ0 commented:
The funny thing is, that 3.6 and 3.7 were not in our build workflow anymore. This looks like a github actions issue and might be temporary. I'll ignore it.

Nevertheless I removed EOL 3.8 and added 3.13 to that workflow in https://github.com/joalla/discogs_client/pull/165/commits/f0d3af2794fb15b2142851fce4874674618ac4c4
